passes/cmds/wrapcell.cc

```diff
@@ -114,43 +187,88 @@ struct WrapcellPass : Pass {
+				for (auto it = str.begin(); it != str.end();) {
+					auto sep = it;
+					for (; sep != str.end() && *sep != ' '; sep++);
+					unused.add(SigBit(wire, std::stoi(std::string(it, sep))));
+					for (it = sep; it != str.end() && *it == ' '; it++);
+				}
+			}
+
 			for (auto cell : module->selected_cells()) {
-				std::optional<std::string> unescaped_name = format(name_fmt, cell->parameters);
+				Module *subm;
+				Cell *subcell;
+
+				if (!ct.cell_known(cell->type))
+					log_error("Non-internal cell type '%s' on cell '%s' in module '%s' unsupported\n",
+							  log_id(cell->type), log_id(cell), log_id(module));
+
+				std::vector<std::pair<IdString, int>> unused_outputs, used_outputs;
+				for (auto conn : cell->connections()) {
+					if (ct.cell_output(cell->type, conn.first))
+					for (int i = 0; i < conn.second.size(); i++) {
+						if (tracking_unused && unused.check(conn.second[i]))
+							unused_outputs.emplace_back(conn.first, i);
+						else
+							used_outputs.emplace_back(conn.first, i);
+					}
+				}
+
+				ContextData context;
+				if (!unused_outputs.empty()) {
+					context.unused_outputs += "_unused";
+					for (auto chunk : collect_chunks(unused_outputs))
+						context.unused_outputs += "_" + RTLIL::unescape_id(chunk.format(cell));
+				}
+
+				std::optional<std::string> unescaped_name = format(name_fmt, cell->parameters, context);
 				if (!unescaped_name)
 					log_error("Formatting error when processing cell '%s' in module '%s'\n",
 							  log_id(cell), log_id(module));
 
 				IdString name = RTLIL::escape_id(unescaped_name.value());
+				if (d->module(name))
+					goto replace_cell;
 
-				if (d->module(name)) {
-					cell->type = name;
-					cell->parameters.clear();
-					continue;
+				subm = d->addModule(name);
+				subcell = subm->addCell("$1", cell->type);
+				for (auto conn : cell->connections()) {
+					if (ct.cell_output(cell->type, conn.first)) {
+						subcell->setPort(conn.first, SigSpec(RTLIL::Sm, conn.second.size()));
+					} else {
+						Wire *w = subm->addWire(conn.first, conn.second.size());
+						w->port_input = true;
+						subcell->setPort(conn.first, w);
+					}
 				}
 
-				if (!ct.cell_known(cell->type))
-					log_error("Non-internal cell type '%s' on cell '%s' in module '%s' unsupported\n",
-							  log_id(cell->type), log_id(cell), log_id(module));
+				for (auto chunk : collect_chunks(used_outputs)) {
+					Wire *w = subm->addWire(chunk.format(cell), chunk.len);
+					w->port_output = true;
+					subcell->connections_[chunk.port].replace(chunk.base, w);
+				}
 
-				Module *subm = d->addModule(name);
-				Cell *subcell = subm->addCell("$1", cell->type);
-				for (auto conn : cell->connections()) {
-					Wire *w = subm->addWire(conn.first, conn.second.size());
-					if (ct.cell_output(cell->type, w->name))
-						w->port_output = true;
-					else
-						w->port_input = true;
-					subcell->setPort(conn.first, w);
+				for (auto chunk : collect_chunks(unused_outputs)) {
```

Comment:
Interestingly, this isn't covered by the test, since commenting out this for loop doesn't make the test fail

Reply:
Let's add `check -assert` which should catch removing this loop

Reply:
No, neither the `check` command nor the internal RTLIL checker catch a marker bit on an output connection:
```
attribute \bar "w3"
attribute \foo 1
module \OR_2_3_3_unused_Y[1]

  wire output 4 \Y[2]

  wire output 3 \Y[0]

  wire width 3 input 2 \B

  wire width 2 input 1 \A

  cell $or $1
    parameter \A_SIGNED 0
    parameter \A_WIDTH 2
    parameter \B_SIGNED 0
    parameter \B_WIDTH 3
    parameter \Y_WIDTH 3
    connect \Y { \Y[2] 1'm \Y[0] }
    connect \B \B
    connect \A \A
  end
end
```
That's something to fix outside of this PR.